enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,1012 @@
+  - fd10:128:20::8/64 
+```
+
+We also propose a new attribute to be added to the network selection elements,
+allowing the workload controller (KubeVirt, which templates the pod) to
+indicate to the CNI plugin which `IPAMClaim` it should use. Remember that IPAM
+CNI will store the generated IP there, while the CNI plugin will consume and
+honor it.
+
+From OVN-Kubernetes perspective, it **must**:
+- **not** delete `IPAMClaim`s for encapsulating entities once their pods
+  are deleted. We need to follow the lifecycle of the encapsulating entity,
+  described via the `metadata.ownerReferences` attribute. I.e. if we're
+  deleting a pod where a KubeVirt VM will run, do not remove the `IPAMClaim`
+  allocation when processing the CNI delete.
+- the IPAM side of OVN-Kubernetes needs to react to the `IPAMClaim` creation,
+  generate an IP for it, and update the `IPAMClaim` status with those IPs.
+- when persistent IPs are requested by the network admin (via the network
+  attachment definition) the CNI side of OVN-Kubernetes must wait for the
+  corresponding `IPAMClaim` to have IPs in its status. Once it does, it will
+  configure the interface.
+
+Deleting the  `IPAMClaim` from the datastore when the pod is deleted will be
+performed by Kubernetes (garbage collection) once the Virtual Machine
+(i.e. owner) is removed. In the example above, that `IPAMClaim` will be
+garbage collected by Kubernetes when the `VirtualMachine` named `vm-a` with UID
+`a0790345-4e84-4257-837a-e3d762d191ab` is deleted.
+
+To support hot-unplug scenarios (from Virtual Machines) the KubeVirt controller
+will need to remove the associated `IPAMClaim` objects. This should only happen
+if the NetworkAttachmentDefinition has the persistent IPs knob enabled.
+
+### Implementation Details/Notes/Constraints
+
+### Risks and Mitigations
+
+When building the OVN-Kubernetes IPAM module IP pools, it will be needed to
+take into account these persistent IPs as well. Thus, there will be a resource
+to sync when adding a pod requesting an attachment to a network allowing
+persistent IPs.
+
+Using informers to know the `IPAMClaim`s CRs will mitigate this risk, since
+we will not actively synchronously call the Kubernetes API for reading.
+
+### Drawbacks
+
+## Design Details
+
+### KubeVirt related changes
+
+#### Pod templating
+
+KubeVirt would need to template the pod accordingly - i.e. the
+`NetworkSelectionElements` must feature the `ipam-claim-reference` attribute to
+indicate to the IPAM CNI / CNI which `IPAMClaim` they should use to persist the
+IP allocation across reboots / migrations.
+
+Check the following network selection elements annotation from a KubeVirt
+launcher pod to see an example:
+
+```bash
+kubectl get pods <launcher pod name> -ojsonpath="{ @.metadata.annotations.k8s\.v1\.cni\.cncf\.io\/networks }" | jq
+[
+  {
+    "name": "tenantred",
+    "namespace": "default",
+    "interface": "pod16367aacb67",
+    "ipam-claim-reference": "vm-a.tenantred-attachment"
```

Comment:
I think would be interesting to take a look at the existing `ips` field in the NetworkSelectionElement, interactions between the two and to check if they are defined in a way that their purpose can be easily told apart.

Reply:
OK. I guess this should be covered in the IPAMClaim specification.